src/core/net.c

```diff
@@ -248,7 +248,9 @@ JANET_NO_RETURN static void janet_sched_accept(JanetStream *stream, JanetFunctio
 /* Adress info */
 
 static int janet_get_sockettype(Janet *argv, int32_t argc, int32_t n) {
-    JanetKeyword stype = janet_optkeyword(argv, argc, n, NULL);
+    JanetKeyword stype = NULL;
+    if(janet_checktype(argv[n], JANET_KEYWORD))
+        stype = janet_optkeyword(argv, argc, n, NULL);
```

Comment:
This change doesn't seem to do anything afaik, since the logic is what optkeyword already does. What is the reason for it?

Reply:
The issue was that when I use `janet_optkeyword` against a non-keyword type, it gives an error at runtime. My understanding is that `janet_optkeyword` errors if what it tries to get is not either nil or a keyword. The way `(net/connect)` is now written it either `:stream` or a `bindhost` (IP) can be in position 2, and a bind host (string) does not count as a keyword, so `janet_optkeyword` errors out.

Reply:
Ok, makes sense. Still, seems like we would have not as good error detection in the non-bind case (say a user passed in a string by accident to `net/listen`.

Reply:
If they pass in a string say "google.com" or "gobledegook", for the former they'll get the same as the latter (unless they are running at an address that resolves to google.com) which would supposedly be a `janet_panic("could not bind outgoing address)`

For listeners, the arity is still the same as it was before and the call never sets "is_bind" so they supposedly could never see it.

Am I missing an error case for when args > 3  && args <= 4?

Reply:
Actually I see what you mean, the error they get would be later along the line as a "cannot bind" error for `(net/connect)` and for listen it should never be anything but an "expected socket type" error. I'm not sure how to fix this and maintain bind support exactly, except maybe to pass down "is_bind" and if so check with `janet_checktype` but otherwise assume and eventually get to the error?

Reply:
Thinking further, with how `janet_get_sockettype` is done, depending on argv, I'm unsure it's possible to get it to check and get a socket type in the case where one is binding. It needs to be pointed at a blank argv to assume "not here, do default" and there's no good way to give it indication "not here, go default" without basically moving it up into `(net/connect)` for the things that `(net/connect)` does. Which ultimately gets you the same error, the one in `(net/connect)` instead of the one in `janet_get_sockettype`.  The only thing it can do is check that argv[2] is correctly set when params length is 4, when params length is 3 there's just no way for it to know if socket type or bindhost was meant to be used...

I'll give it more thought, but I'm not sure I have any great ideas on how to keep socket type errors only in `janet_get_sockettype`, so if you or someone else has a good idea I'm all ears.

Reply:
I think there is some misunderstanding here - get_sockettype is just a simple function to read `:stream`, `:datagram`, or `nil` from the arglist. By removing the check, you are allowing errors through in other functions that use it.

Reply:
For example, `(net/listen "127.0.0.1" "8000" @"blahblahblahblah")` should throw an error - with this change, it will continue as normal.

Reply:
You're correct, and it sucks, but I'm gonna fix that here in a bit. Basically, gonna move that check out of the socket type getter and into net/connect (gonna try to collect all address infos at the head of the function in net/connect with a strict socket type check when argc == 4 and a lax (caller checks) check when argc == 3). The result will be that net/connect could error in socket type, or it can error in a bind failure.

Reply:
I guess we are used to using `nil` for the unused argument. For example in `fiber/new` when you are not providing the resume value, but you provide a supervisor channel

Reply:
Pushed fix, keeps other `net/*` from being effected by changes to `net/connect`. The issue here is that `net/connect` can still have an issue like how you mentioned with `net/listen` (except that unless the user managed to pick something hostname or ip address like, they'd get a bind error rather than an error about calling the function with the wrong types).

The only way to fix that would be to break the `net/connect` API call and have it take keyword arguments for all optional arguments instead of variadic limited arity arguments. If you have any ideas on how to detect nil in position 3 or 4 when the argc can be 3 or 4 let me know but without a language change (sparse argv) or API change I'm not really sure how to make `net/connect` for sure give an argument error for not using keyword when it should be a keyword instead of a bind error.

I could leave a note on the function to "break API and use keyword arguments in $version" where I would guess version would be 2.0.0 (I don't think breaking api changes are wanted for the second number (like 0.X.0).

The final idea I have is perhaps to create a strict 4 argument `net/connect_impl` and then create functions in the boot.janet that wrap around it that support the old way (like how the code is now) and a new way later on down the road (I don't think janet allows a function to have multiple definitions dependent upon argument types where it'd need to at runtime pick a call to one that takes optionals and another that takes keywords excepting if we use a function intermediate to dispatch to the correctly defined underlying function).

Reply:
> if one can create a function with arguments like [& xs] where it can be called like it's using keyword arguments...

Note sure I follow, but do you mean like this?
```
(defn make-recipe-3
  "Create some kind of recipe using &keys..."
  [&keys {:flour flour
          :sugar sugar
          :baking-soda soda
          :water water
          :eggs eggs
          :vanilla-extract vanilla}]
  (def dry [flour sugar soda])
  (def wet [water eggs vanilla])
  {:name "underspecified-cake" :wet wet :dry dry})

# Calling this function is a bit different now - no struct
(make-recipe-3
  :flour 1
  :sugar 1
  :baking-soda 0.5
  :water 2
  :eggs 2
  :vanilla-extract 0.5)
```
via: https://janet-lang.org/docs/functions.html#Keyword-style-arguments

Reply:
@sogaiu 

Say we have a cfunction `net/connect_impl (host port sockettype bind)` (no opt args)

Now say we create in boot.janet a dispatcher that will use that:

`(net/connect [host port [& xs]])`

Can I use xs like keyword arguments? The idea is to allow a call like:
`(net/connect [host port &opt sockettype bind]` 
and at the same time allow 
`(net/connect [host port [&keys {:type type :bind bind}]])`
which would dispatch to `net/connect_impl`

Which the first one defined (`(net/connect [host port [& xs]])`) would dispatch to depending on what sorts of args it got.

Reply:
I don't think so, but am not sure.

May be @bakpakin or @andrewchambers know better.

Reply:
Just realized, but are keyword arguments even optional? I mean

```
(net/connect host port :type :stream :bind "127.0.0.1")
# should be the same as below
(net/connect host port :bind "127.0.0.1")
# right?
```

Which brings me to thinking it'd be nice to have keys that have a default set in the function itself... such that you could do something like:

```
(net/connect [host port [&keys {:type &default stream :bind &default ""}]])
```

Tho I guess in all reality... a really well done macro could possibly do all the hard work there. Though I don't think a macro could tell the difference between "they meant to set socket type with :stream but gave a string" and "they actually meant it as an IP so it's ok" either...

Reply:
Fixed to work as @bakpakin specified in [#789](https://github.com/janet-lang/janet/pull/789#issuecomment-914571833)